lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+        PCH->setPC(Builder, NewPC);
+      }
+    }
+  }
+
+  // Some LLVM passes used later in the pipeline scan for cut-offs, meaning that
+  // further computation may not be done when they are reached; making some
+  // optimizations opportunities missed. Hence, we set the involved thresholds
+  // (e.g., the maximum value that MemorySSA uses to take into account
+  // stores/phis) to have initial unbounded value.
+  auto getUOpt = getOption<unsigned>;
+  StringMap<cl::Option *> &Options = cl::getRegisteredOptions();
+  static constexpr const char *MemSSALimit = "memssa-check-limit";
+  static constexpr const char *MemDepBlockLimit = "memdep-block-scan-limit";
+
+  auto MemSSALimitDefault = getUOpt(Options, MemSSALimit)->getValue();
+  auto MemDepBlockLimitDefault = getUOpt(Options, MemDepBlockLimit)->getValue();
+
+  getUOpt(Options, MemSSALimit)->setInitialValue(UINT_MAX);
+  getUOpt(Options, MemDepBlockLimit)->setInitialValue(UINT_MAX);
+
+  // Run optimization passes over the disposable function.
+  {
+    FunctionPassManager FPM;
+
+    // First stage: simplify the IR, promote the CSVs to local variables,
+    // compute subexpressions elimination and resolve redundant expressions in
+    // order to compute the stack height.
+    FPM.addPass(RemoveNewPCCallsPass());
+    FPM.addPass(RemoveHelperCallsPass());
+    FPM.addPass(PromoteGlobalToLocalPass());
+    FPM.addPass(SimplifyCFGPass());
+    FPM.addPass(SROA());
+    FPM.addPass(EarlyCSEPass(true));
+    FPM.addPass(JumpThreadingPass());
+    FPM.addPass(UnreachableBlockElimPass());
+    FPM.addPass(InstCombinePass(true));
+    FPM.addPass(EarlyCSEPass(true));
+    FPM.addPass(SimplifyCFGPass());
+    FPM.addPass(MergedLoadStoreMotionPass());
+    FPM.addPass(GVN());
+
+    // Second stage: add alias analysis info and canonicalize i2p + add into
+    // geps. Since the IR may change remarkably, another round of passes is
+    // necessary to take more optimization opportunities.
+    FPM.addPass(SegregateDirectStackAccessesPass());
+    FPM.addPass(EarlyCSEPass(true));
+    FPM.addPass(InstCombinePass(true));
+    FPM.addPass(GVN());
+
+    // Third stage: if enabled, serialize the results and dump the functions on
+    // disk with the alias information included as comments.
+    if (IndirectBranchInfoSummaryPath.getNumOccurrences() == 1)
+      FPM.addPass(IndirectBranchInfoPrinterPass(OutputIBI));
+
+    if (AAWriterPath.getNumOccurrences() == 1)
+      FPM.addPass(AAWriterPass(OutputAAWriter));
+
+    FunctionAnalysisManager FAM;
+    FAM.registerPass([] {
+      AAManager AA;
+      AA.registerFunctionAnalysis<BasicAA>();
+      AA.registerFunctionAnalysis<ScopedNoAliasAA>();
+
+      return AA;
+    });
+
+    ModuleAnalysisManager MAM;
+    MAM.registerPass([&] { return GeneratedCodeBasicInfoAnalysis(); });
+    FAM.registerPass([&] { return GeneratedCodeBasicInfoAnalysis(); });
+    FAM.registerPass([&] { return ModuleAnalysisManagerFunctionProxy(MAM); });
```

Comment:
Move these close to the declaration of `FAM`

Reply:
Uhmm, I've always seen that pattern (i.e., `ModuleAnalysisManagerFunctionProxy` registered after MAM declaration). Let's leave the pipeline and its function/module analysis managers as they are in order not break anything; and come back on this in the future.

Reply:
`FAM` and `MAM` do not know about each other yet until the last line, the order in which you `registerPass` before that cannot matter. In general, I don't think the order in which you call `registerPass` matters.
Let's not be overly superstitious here.

```c++
    ModuleAnalysisManager MAM;
    MAM.registerPass([&] { return GeneratedCodeBasicInfoAnalysis(); });

    FunctionAnalysisManager FAM;
    FAM.registerPass([] {
      AAManager AA;
      AA.registerFunctionAnalysis<BasicAA>();
      AA.registerFunctionAnalysis<ScopedNoAliasAA>();

      return AA;
    });
    FAM.registerPass([&] { return GeneratedCodeBasicInfoAnalysis(); });
    FAM.registerPass([&] { return ModuleAnalysisManagerFunctionProxy(MAM); });
```

Reply:
OK!